### ISSUE TITLE
Make RulePolicy more efficient with large trackers

### ISSUE THREAD

**Issue by desmarchris** [RasaHQ/rasa#9089]
**Description of Problem**:
Large trackers can cause a slowdown / memory spike when making a prediction with RulePolicy. 

**Definition of Done**:
- [x] Reproduce problem
- [ ] Scope possible fixes

**Comment by desmarchris** [comment]
@TyDunn I wasn't sure what squad this would belong under, lmk if it's not enable :)

**Comment by TyDunn** [comment]
@desmarchris I'd check out [this thread on Slack](https://rasa-hq.slack.com/archives/C01HHMR4X8S/p1625824687082600) that discusses some related things

**Comment by kedz** [comment]
@JEM-Mosig does your bug fix [PR](https://github.com/RasaHQ/rasa/pull/9370) possibly also address this issue?

**Comment by JEM-Mosig** [comment]
@kedz No, this will not affect RulePolicy, since I only touch the AugmentedMemoizationPolicy class and RulePolicy is derived from MemoizationPolicy.

**Comment by JEM-Mosig** [comment]
I'd rephrase this issue to "make RulePolicy faster with large trackers"

**Comment by desmarchris** [comment]
changed it slightly bc it's not just about time but also memory/cpu consumption

**Comment by JEM-Mosig** [comment]
@desmarchris I'm trying to reproduce this right now. To be sure: the problem was with _inference_ only, right? Not with training. Do we have an example tracker?

**Comment by desmarchris** [comment]
correct, just at inference. I'm sending you an example for a single prediction that's 2500 lines long on Slack.

**Comment by JEM-Mosig** [comment]
Oh, I think the linear time increase is actually the issue. Mapping policy didn't check the tracker at all (except last event), so it'd be constant time for that one.

**Comment by desmarchris** [comment]
the specific issue at hand here was literally printing 2500 lines to make a prediction caused a CPU spike that brought the pod close to its max capacity

**Comment by JEM-Mosig** [comment]
_Printing_? So this has nothing to do with the policy, but only with the fact that debug was on?

**Comment by desmarchris** [comment]
I think so? but debug was needed to help find what went wrong with the bot. I don't know anyone running a bot in prod without debug turned on

**Comment by JEM-Mosig** [comment]
> I don't know anyone running a bot in prod without debug turned on
That is very concerning. It's not meant to be used this way. I suppose we can have RulePolicy just not print the state in the short term, and in the long term... find a way to store debug logs more quickly? @koaning had this idea of a streamlined debug log (`--insights`?), but I suppose customers would just always go for printing everything. So we'd have to add a "secret debug flag" that is really only to be used for us inside Rasa.

**Comment by JEM-Mosig** [comment]
I see we actually have a Issue https://github.com/RasaHQ/rasa/issues/7295 for that already.

I'll close the present issue now, noting that the abovementioned issue would fix the slow debug output, which caused the present issue in the first place. 

If we wanted to actually accelerate `RulePolicy`, we could speed up the lookup of rules. What makes this slightly complicated is that in contrast to "MappingPolicy rules", the "RulePolicy rules" can be nested. I.e. you can have a short rule inside a long rule. Even if the short rule applies, you still need to check if you can find a long rule that is more specific. Surely we can create a custom algorithm that is faster than the linear time one that we have right now. Again, given that this doesn't actually cause an issue with the customer, and the present issue is about scoping solutions, I'll close the present issue.
